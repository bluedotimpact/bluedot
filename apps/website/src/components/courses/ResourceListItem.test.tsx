import React from 'react';
import { render } from '@testing-library/react';
import {
  describe, expect, it, vi,
} from 'vitest';
import { ResourceListItem } from './ResourceListItem';

// Mock next-auth
vi.mock('next-auth/react', () => ({
  useSession: () => ({
    data: {
      user: {
        email: 'test@example.com',
        name: 'Test User',
      },
    },
    status: 'authenticated',
  }),
}));

// Mock next/router
vi.mock('next/router', () => ({
  useRouter: () => ({
    query: { courseSlug: 'test-course', unitNumber: '1' },
    pathname: '/courses/[courseSlug]/units/[unitNumber]',
    push: vi.fn(),
    replace: vi.fn(),
  }),
}));

// Mock FaviconImage component
vi.mock('./FaviconImage', () => ({
  FaviconImage: ({ url, alt }: { url: string; alt: string }) => (
    <img src={url} alt={alt} />
  ),
}));

// Mock MarkdownExtendedRenderer
vi.mock('./MarkdownExtendedRenderer', () => ({
  default: function MarkdownExtendedRenderer({ children, className }: { children: React.ReactNode; className?: string }) {
    return <div className={className}>{children}</div>;
  },
}));

describe('ResourceListItem - Listen to Article Feature', () => {
  const baseResource = {
    id: 'test-resource-1',
    resourceName: 'Introduction to AI Safety',
    resourceType: 'article',
    resourceLink: 'https://example.com/article',
    resourceGuide: 'This is a guide to the resource',
    authors: 'John Doe',
    timeFocusOnMins: 10,
    coreFurtherMaybe: null,
    readingOrder: null,
    unitId: 'unit123',
    avgRating: null,
    syncedAudioUrl: null,
    year: 2024,
    autoNumberId: null,
  };

  it('should render metadata without Listen to article button when no audio URL', () => {
    const { container, queryByText } = render(
      <ResourceListItem resource={baseResource} />,
    );

    // Should show author, year, and time
    expect(queryByText(/John Doe/)).toBeTruthy();
    expect(queryByText(/2024/)).toBeTruthy();
    expect(queryByText(/10 min/)).toBeTruthy();

    // Should NOT show Listen to article
    expect(queryByText('Listen to article')).toBeFalsy();

    expect(container.querySelector('.resource-item__bottom-metadata')).toMatchSnapshot();
  });

  it('should render Listen to article button when audio URL is provided', () => {
    const resourceWithAudio = {
      ...baseResource,
      syncedAudioUrl: 'https://open.spotify.com/episode/abc123',
    };

    const { container, getByText } = render(
      <ResourceListItem resource={resourceWithAudio} />,
    );

    // Should show author, time, and Listen to article
    expect(getByText(/John Doe/)).toBeTruthy();
    expect(getByText(/10 min/)).toBeTruthy();
    expect(getByText('Listen to article')).toBeTruthy();

    expect(container.querySelector('.resource-item__bottom-metadata')).toMatchSnapshot();
  });

  it('should render Listen to article with proper separator when metadata exists', () => {
    const resourceWithAudio = {
      ...baseResource,
      authors: 'Jane Smith',
      timeFocusOnMins: 15,
      syncedAudioUrl: 'https://open.spotify.com/episode/xyz789',
    };

    const { container } = render(
      <ResourceListItem resource={resourceWithAudio} />,
    );

    const metadata = container.querySelector('.resource-item__bottom-metadata');
    expect(metadata).toMatchSnapshot();

    // Check that separators are rendered
    const textContent = metadata?.textContent;
    expect(textContent).toContain('Jane Smith');
    expect(textContent).toContain('路');
    expect(textContent).toContain('15 min');
    expect(textContent).toContain('Listen to article');
  });

  it('should render Listen to article even without other metadata', () => {
    const resourceWithOnlyAudio = {
      ...baseResource,
      authors: null,
      timeFocusOnMins: null,
      syncedAudioUrl: 'https://open.spotify.com/episode/solo123',
    };

    const { container, getByText } = render(
      <ResourceListItem resource={resourceWithOnlyAudio} />,
    );

    // Should show Listen to article button even when no other metadata
    expect(getByText('Listen to article')).toBeTruthy();

    // Should show year and separator before audio button
    const metadata = container.querySelector('.resource-item__bottom-metadata');
    expect(metadata?.textContent).toContain('2024');
    expect(metadata?.textContent).toContain('路');
    expect(metadata?.textContent).toContain('Listen to article');

    expect(metadata).toMatchSnapshot();
  });

  it('should handle various metadata combinations with audio URL', () => {
    // Only author and audio
    const authorAndAudio = {
      ...baseResource,
      authors: 'Alice Brown',
      timeFocusOnMins: null,
      syncedAudioUrl: 'https://open.spotify.com/episode/author123',
    };

    const { container: container1 } = render(
      <ResourceListItem resource={authorAndAudio} />,
    );
    expect(container1.querySelector('.resource-item__bottom-metadata')).toMatchSnapshot();

    // Only time and audio
    const timeAndAudio = {
      ...baseResource,
      authors: null,
      timeFocusOnMins: 20,
      syncedAudioUrl: 'https://open.spotify.com/episode/time123',
    };

    const { container: container2 } = render(
      <ResourceListItem resource={timeAndAudio} />,
    );
    expect(container2.querySelector('.resource-item__bottom-metadata')).toMatchSnapshot();
  });

  it('should display year field correctly in metadata', () => {
    const { queryByText } = render(
      <ResourceListItem resource={baseResource} />,
    );

    // Should show year between author and time
    expect(queryByText(/2024/)).toBeTruthy();
  });

  it('should handle metadata without year field', () => {
    const resourceWithoutYear = {
      ...baseResource,
      year: null,
    };

    const { container, queryByText } = render(
      <ResourceListItem resource={resourceWithoutYear} />,
    );

    // Should show author and time but not year
    expect(queryByText(/John Doe/)).toBeTruthy();
    expect(queryByText(/10 min/)).toBeTruthy();
    expect(queryByText(/2024/)).toBeFalsy();

    // Check proper separator handling without year
    const metadata = container.querySelector('.resource-item__bottom-metadata');
    const textContent = metadata?.textContent;
    expect(textContent).toContain('John Doe');
    expect(textContent).toContain('路');
    expect(textContent).toContain('10 min');
  });

  it('should handle only year field in metadata', () => {
    const resourceWithOnlyYear = {
      ...baseResource,
      authors: null,
      timeFocusOnMins: null,
    };

    const { container, queryByText } = render(
      <ResourceListItem resource={resourceWithOnlyYear} />,
    );

    // Should show only year
    expect(queryByText(/2024/)).toBeTruthy();
    expect(queryByText(/John Doe/)).toBeFalsy();
    expect(queryByText(/10 min/)).toBeFalsy();

    // Should not have separators when only year is present
    const metadata = container.querySelector('.resource-item__bottom-metadata');
    expect(metadata?.textContent).not.toContain('路');
  });
});
