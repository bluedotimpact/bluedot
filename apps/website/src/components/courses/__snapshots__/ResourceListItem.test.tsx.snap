// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceListItem - Listen to Article Feature > should handle various metadata combinations with audio URL 1`] = `
<div
  class="resource-item__bottom-metadata mt-4 flex flex-wrap items-center gap-x-1 gap-y-2"
>
  <p
    class="bluedot-p not-prose text-gray-600 text-[13px] font-medium leading-[140%] tracking-[-0.005em]"
  >
    <span>
      Alice Brown
    </span>
    <span>
       · 
    </span>
    <span>
      2024
    </span>
    <span>
       ·
    </span>
  </p>
  <button
    aria-label="Listen to article: Introduction to AI Safety (opens in Spotify)"
    class="flex flex-row items-center gap-1.5 h-[18px] group cursor-pointer transition-colors duration-200 bg-transparent border-none p-0"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        class="group-hover:stroke-[#13132E] transition-colors duration-200"
        height="14.75"
        rx="7.375"
        stroke="#6A6F7A"
        stroke-width="1.25"
        width="14.75"
        x="0.625"
        y="0.625"
      />
      <path
        class="group-hover:fill-[#13132E] transition-colors duration-200"
        d="M6 10.4839V5.51675C6 5.13551 6.40956 4.89452 6.74282 5.07967L11.2133 7.56325C11.5562 7.75375 11.5562 8.2469 11.2133 8.4374L6.74282 10.921C6.40956 11.1061 6 10.8651 6 10.4839Z"
        fill="#6A6F7A"
      />
    </svg>
    <span
      class="text-[13px] font-medium leading-[140%] tracking-[-0.005em] text-[#6A6F7A] group-hover:text-[#13132E] transition-colors duration-200"
    >
      Listen to article
    </span>
  </button>
</div>
`;

exports[`ResourceListItem - Listen to Article Feature > should handle various metadata combinations with audio URL 2`] = `
<div
  class="resource-item__bottom-metadata mt-4 flex flex-wrap items-center gap-x-1 gap-y-2"
>
  <p
    class="bluedot-p not-prose text-gray-600 text-[13px] font-medium leading-[140%] tracking-[-0.005em]"
  >
    <span>
      2024
    </span>
    <span>
       · 
    </span>
    <span>
      20
       min
    </span>
    <span>
       ·
    </span>
  </p>
  <button
    aria-label="Listen to article: Introduction to AI Safety (opens in Spotify)"
    class="flex flex-row items-center gap-1.5 h-[18px] group cursor-pointer transition-colors duration-200 bg-transparent border-none p-0"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        class="group-hover:stroke-[#13132E] transition-colors duration-200"
        height="14.75"
        rx="7.375"
        stroke="#6A6F7A"
        stroke-width="1.25"
        width="14.75"
        x="0.625"
        y="0.625"
      />
      <path
        class="group-hover:fill-[#13132E] transition-colors duration-200"
        d="M6 10.4839V5.51675C6 5.13551 6.40956 4.89452 6.74282 5.07967L11.2133 7.56325C11.5562 7.75375 11.5562 8.2469 11.2133 8.4374L6.74282 10.921C6.40956 11.1061 6 10.8651 6 10.4839Z"
        fill="#6A6F7A"
      />
    </svg>
    <span
      class="text-[13px] font-medium leading-[140%] tracking-[-0.005em] text-[#6A6F7A] group-hover:text-[#13132E] transition-colors duration-200"
    >
      Listen to article
    </span>
  </button>
</div>
`;

exports[`ResourceListItem - Listen to Article Feature > should render Listen to article button when audio URL is provided 1`] = `
<div
  class="resource-item__bottom-metadata mt-4 flex flex-wrap items-center gap-x-1 gap-y-2"
>
  <p
    class="bluedot-p not-prose text-gray-600 text-[13px] font-medium leading-[140%] tracking-[-0.005em]"
  >
    <span>
      John Doe
    </span>
    <span>
       · 
    </span>
    <span>
      2024
    </span>
    <span>
       · 
    </span>
    <span>
      10
       min
    </span>
    <span>
       ·
    </span>
  </p>
  <button
    aria-label="Listen to article: Introduction to AI Safety (opens in Spotify)"
    class="flex flex-row items-center gap-1.5 h-[18px] group cursor-pointer transition-colors duration-200 bg-transparent border-none p-0"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        class="group-hover:stroke-[#13132E] transition-colors duration-200"
        height="14.75"
        rx="7.375"
        stroke="#6A6F7A"
        stroke-width="1.25"
        width="14.75"
        x="0.625"
        y="0.625"
      />
      <path
        class="group-hover:fill-[#13132E] transition-colors duration-200"
        d="M6 10.4839V5.51675C6 5.13551 6.40956 4.89452 6.74282 5.07967L11.2133 7.56325C11.5562 7.75375 11.5562 8.2469 11.2133 8.4374L6.74282 10.921C6.40956 11.1061 6 10.8651 6 10.4839Z"
        fill="#6A6F7A"
      />
    </svg>
    <span
      class="text-[13px] font-medium leading-[140%] tracking-[-0.005em] text-[#6A6F7A] group-hover:text-[#13132E] transition-colors duration-200"
    >
      Listen to article
    </span>
  </button>
</div>
`;

exports[`ResourceListItem - Listen to Article Feature > should render Listen to article even without other metadata 1`] = `
<div
  class="resource-item__bottom-metadata mt-4 flex flex-wrap items-center gap-x-1 gap-y-2"
>
  <p
    class="bluedot-p not-prose text-gray-600 text-[13px] font-medium leading-[140%] tracking-[-0.005em]"
  >
    <span>
      2024
    </span>
    <span>
       ·
    </span>
  </p>
  <button
    aria-label="Listen to article: Introduction to AI Safety (opens in Spotify)"
    class="flex flex-row items-center gap-1.5 h-[18px] group cursor-pointer transition-colors duration-200 bg-transparent border-none p-0"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        class="group-hover:stroke-[#13132E] transition-colors duration-200"
        height="14.75"
        rx="7.375"
        stroke="#6A6F7A"
        stroke-width="1.25"
        width="14.75"
        x="0.625"
        y="0.625"
      />
      <path
        class="group-hover:fill-[#13132E] transition-colors duration-200"
        d="M6 10.4839V5.51675C6 5.13551 6.40956 4.89452 6.74282 5.07967L11.2133 7.56325C11.5562 7.75375 11.5562 8.2469 11.2133 8.4374L6.74282 10.921C6.40956 11.1061 6 10.8651 6 10.4839Z"
        fill="#6A6F7A"
      />
    </svg>
    <span
      class="text-[13px] font-medium leading-[140%] tracking-[-0.005em] text-[#6A6F7A] group-hover:text-[#13132E] transition-colors duration-200"
    >
      Listen to article
    </span>
  </button>
</div>
`;

exports[`ResourceListItem - Listen to Article Feature > should render Listen to article with proper separator when metadata exists 1`] = `
<div
  class="resource-item__bottom-metadata mt-4 flex flex-wrap items-center gap-x-1 gap-y-2"
>
  <p
    class="bluedot-p not-prose text-gray-600 text-[13px] font-medium leading-[140%] tracking-[-0.005em]"
  >
    <span>
      Jane Smith
    </span>
    <span>
       · 
    </span>
    <span>
      2024
    </span>
    <span>
       · 
    </span>
    <span>
      15
       min
    </span>
    <span>
       ·
    </span>
  </p>
  <button
    aria-label="Listen to article: Introduction to AI Safety (opens in Spotify)"
    class="flex flex-row items-center gap-1.5 h-[18px] group cursor-pointer transition-colors duration-200 bg-transparent border-none p-0"
    type="button"
  >
    <svg
      aria-hidden="true"
      fill="none"
      height="16"
      viewBox="0 0 16 16"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        class="group-hover:stroke-[#13132E] transition-colors duration-200"
        height="14.75"
        rx="7.375"
        stroke="#6A6F7A"
        stroke-width="1.25"
        width="14.75"
        x="0.625"
        y="0.625"
      />
      <path
        class="group-hover:fill-[#13132E] transition-colors duration-200"
        d="M6 10.4839V5.51675C6 5.13551 6.40956 4.89452 6.74282 5.07967L11.2133 7.56325C11.5562 7.75375 11.5562 8.2469 11.2133 8.4374L6.74282 10.921C6.40956 11.1061 6 10.8651 6 10.4839Z"
        fill="#6A6F7A"
      />
    </svg>
    <span
      class="text-[13px] font-medium leading-[140%] tracking-[-0.005em] text-[#6A6F7A] group-hover:text-[#13132E] transition-colors duration-200"
    >
      Listen to article
    </span>
  </button>
</div>
`;

exports[`ResourceListItem - Listen to Article Feature > should render metadata without Listen to article button when no audio URL 1`] = `
<div
  class="resource-item__bottom-metadata mt-4 flex flex-wrap items-center gap-x-1 gap-y-2"
>
  <p
    class="bluedot-p not-prose text-gray-600 text-[13px] font-medium leading-[140%] tracking-[-0.005em]"
  >
    <span>
      John Doe
    </span>
    <span>
       · 
    </span>
    <span>
      2024
    </span>
    <span>
       · 
    </span>
    <span>
      10
       min
    </span>
  </p>
</div>
`;
